### DISCUSSION TITLE
Free energy minimization doest not end at the minimum

### DISCUSSION THREAD

**Discussion by mahelz** [SSCHAcode/python-sscha#176] [Q&A]
Dear Developers of SSCHA,

I try to do the minimization of the free energy using an ab-initio tight binding model called Fireball to compute forces and energies. However, the minimization does not end at the minimum of the free energy.

I tried to follow the step of tutorial 2 to submit DFT calculations manually on a cluster. I make sure to convert the forces and energies in the appropriate units: energies (Ry) and forces (Ry/Bohr) . I tried to use more configurations for each population (from 20 to 100 configurations), I also tried a denser q-mesh for the dynamical matrices (4x4) to (8x8), but I always get the same results with the minimization ending not at the minimum of the free energy.

Here is my input file to generate an ensemble:

```
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble
import numpy as np

dyn = CC.Phonons.Phonons("dyn_pop11_", nqirr=4)
dyn.Symmetrize()
dyn.ForcePositiveDefinite()

temperature = 0 # 300 K
ensemble = sscha.Ensemble.Ensemble(dyn, temperature)

number_of_configurations = 20
ensemble.generate(number_of_configurations)

save_directory = "pop12_data"
population_id = 12
ensemble.save(save_directory, population_id)

```

Then I do my DFT calculation and I start the minimization with the following script:

```
import cellconstructor as CC, cellconstructor.Phonons
import sscha, sscha.Ensemble, sscha.SchaMinimizer, sscha.Utilities

dyn = CC.Phonons.Phonons("dyn_start_population12_", 4)

ensemble = sscha.Ensemble.Ensemble(dyn,0)
ensemble.load(data_dir="pop12_data", population=12, N=20)

ioinfo = sscha.Utilities.IOInfo()
ioinfo.SetupSaving("minim_pop12")

minim = sscha.SchaMinimizer.SSCHA_Minimizer(ensemble)
minim.init()
minim.meaningful_factor = 0.001
minim.min_step_struc = 0.05
minim.min_step_dyn = 0.002
minim.run(custom_function_post = ioinfo.CFP_SaveAll)
minim.finalize()
minim.dyn.save_qe("dyn_pop12_")
minim.plot_results(save_filename="out_minim_pop12")

```
And here is the kind of data that I get during the minimization:

![Figure_1](https://github.com/SSCHAcode/python-sscha/assets/151047743/76ce4862-d30e-444f-a59f-09e717950c9d)

Please can you tell me where I could look to fix this problem? Thanks in advance.

All the best,
Mahé Lezoualc'h

**Comment by mesonepigreco** [comment]
Hi, are you sure that the figure you get results from just one population (the population 12)? It could be a plotting error if this comes from an old code version. You can install the latest version in a new environment; since you are using the manual minimization, it should work, and compare if it gives the same results.
You are using just 20 configurations in your ensemble. However, the minimization seems to proceed smoothly, 
Can you also attach the data of ``minim_pop12.freqs`` and ``minim_pop12.dat`` generated by the code? In this way, we can zoom in on the region where the free energy is increasing to understand better what is happening.

**Reply by diegomartinez2** [reply]
You are right, the figures do not look to correspond to one population calculation. From the effective sample size figure and the structure gradient, looks like this is a "vc-relax" calculation. Or maybe is the result of several independent calculations put together in graph.

**Comment by mahelz** [comment]
Hi,

Yes exactly, I put all the independent calculation in one graph. Here are the data minim_pop12.freqs and minim_pop12.dat:

![minim_pop12](https://github.com/SSCHAcode/python-sscha/assets/151047743/a3407c13-e3e4-4a8d-ad55-8976487d65a9)
![minim_pop12_freqs](https://github.com/SSCHAcode/python-sscha/assets/151047743/2ff54f5f-eaaf-4af8-b2db-66003e574a14)

I downloaded the latest version of SSCHA and I get similar results. Here is the all minimization put together in one graph:

![minim_tot_latest-v](https://github.com/SSCHAcode/python-sscha/assets/151047743/d15d4922-0af2-45dd-8008-9551e7e7c1bd)

And here the data for the last population:

![minim_pop24](https://github.com/SSCHAcode/python-sscha/assets/151047743/940978b7-249b-43f6-897b-a4741a068c9f)
![minim_pop24_freqs](https://github.com/SSCHAcode/python-sscha/assets/151047743/7cfc1ae5-d2d2-4e9f-a7f7-60dccb8d04b3)

The input script are the same that I used for the previous calculation with the same parameters. I still have the free energy that converges of about +0.01 ev / unit cell of the minimum.

**Reply by mahelz** [reply]
It seems strange to me that the minimization proceeds smoothly for only 20 configurations. I tried with quantum espresso for 20 configurations and the same parameters. Here is what I get (with the old version of SSCHA)

![minim_tot_qe_freqs](https://github.com/SSCHAcode/python-sscha/assets/151047743/1911f25d-5e6c-4db2-93c0-7894742ceecd)
![minim_tot_qe](https://github.com/SSCHAcode/python-sscha/assets/151047743/3bc9b6b5-9449-40a0-9691-89064b17c071)

**Reply by mesonepigreco** [reply]
So the calculation done with quantum espresso seems fine, while the one with your interface to the tight-binding model no. 

Try to test the concistency between energy and forces in eV in the tight-binding model as it is interfaced with the SSCHA, because the model seems to relax very nicely also with the tight-binding, and the raising of the free energy seems most likely a mismatch between units (or sign).

**Comment by mahelz** [comment]
I checked the units while interfacing the tight-binding model to SSCHA, it looks correct. I tried to do the minimization for H3S as given in the tutorials and I don't have the raising of the free energy at the end of the calculation. I used 20, 50 and 100 configurations for this example with the same results. So the problem might come from the system that I'm studying. I am trying to do other tests on other systems.

![H3S_freqs](https://github.com/SSCHAcode/python-sscha/assets/151047743/77b6abbc-ff73-4123-970e-6d35e74c86fd)
![H3S_data](https://github.com/SSCHAcode/python-sscha/assets/151047743/f633ce6b-fe0a-4ba8-9f4a-9b3b67ba40f8)

**Comment by mahelz** [comment]
Hi,

I did some tests on other systems. I always get a strange minimization for system such as TMDs: bulk and monolayer. However for system like h3s or Au fcc, the minimization looks good. Do you think this could come from the calculation of the forces with the tight-binding model ? If the forces are not well calculated, this can happen to have weird behaviour in the minimization of the free energy?
Thanks for your anwers.

Best, 
Mahé

**Reply by mesonepigreco** [reply]
It is very strange. Indeed, it should not happen. 
We have done TMDs with SSCHA and QE without this problem.
I still suspect it is a problem with the coherence between forces and energies, either units, sign, or some problem in the actual calculation of forces, as in that way I could see why you are getting this behavior. This is really delicate, easy to mess, difficult to spot. Have you tested if forces are consistent with a finite difference approach? Is the H3S test you did with the tight-binding model or with QE? If this problem always and only arises when using the tight-binding, then most likely the problem is there.

**Reply by mahelz** [reply]
I did the H3S test with Fireball and QE, both converge well. I didn't try to see if forces are consistent with a finite difference approach. I'm gonna look in more details for the calculation of forces in the code. Thank you for your reply.
